application/src/Form/ResourceBatchUpdateForm.php

```diff
@@ -1,11 +1,329 @@
+            $preData['remove']['o:is_open'] = $data['is_open'];
+        }
+        if (-1 == $data['resource_template']) {
+            $preData['remove']['o:resource_template'] = ['o:id' => null];
+        } elseif (is_numeric($data['resource_template'])) {
+            $preData['remove']['o:resource_template'] = ['o:id' => $data['resource_template']];
+        }
+        if (-1 == $data['resource_class']) {
+            $preData['remove']['o:resource_class'] = ['o:id' => null];
+        } elseif (is_numeric($data['resource_class'])) {
+            $preData['remove']['o:resource_class'] = ['o:id' => $data['resource_class']];
+        }
+        if (isset($data['remove_from_item_set'])) {
+            $preData['remove']['o:item_set'] = $data['remove_from_item_set'];
+        }
+        if (isset($data['clear_property_values'])) {
+            $preData['remove']['clear_property_values'] = $data['clear_property_values'];
+        }
+        if (!empty($data['clear_language'])) {
+            $preData['remove']['o:lang'] = null;
+        }
+        if (!empty($data['language'])) {
+            $preData['remove']['o:lang'] = $data['language'];
+        }
+
+        // Set the data to append.
+        if (!empty($data['value'])) {
+            foreach ($data['value'] as $value) {
+                $valueObj = [
+                    'property_id' => $value['property_id'],
+                    'type' => $value['type'],
+                ];
+                switch ($value['type']) {
+                    case 'uri':
+                        $valueObj['@id'] = $value['id'];
+                        $valueObj['o:label'] = $value['label'];
+                        break;
+                    case 'resource':
+                        $valueObj['value_resource_id'] = $value['value_resource_id'];
+                        break;
+                    case 'literal':
+                    default:
+                        $valueObj['@value'] = $value['value'];
+                }
+                $preData['append'][$value['property_id']][] = $valueObj;
+            }
+        }
+        if (isset($data['add_to_item_set'])) {
+            $preData['append']['o:item_set'] = array_unique($data['add_to_item_set']);
+        }
+
+        // Set remaining elements according to attribute data-collection-action.
+        $processeds = [
+            'is_public', 'is_open', 'resource_template', 'resource_class',
+            'remove_from_item_set', 'add_to_item_set',
+            'clear_property_values', 'value',
+            'clear_language', 'language',
+            'csrf', 'id', 'o:id',
+        ];
+
+        foreach ($data as $key => $value) {
+            if (is_numeric($key) || in_array($key, $processeds)
+                || is_null($value) || $value === ''
+            ) {
+                continue;
+            }
+            $collectionAction = $this->has($key)
+                ? $this->get($key)->getAttribute('data-collection-action')
```

Comment:
I'm not sure of the purpose of `data-collection-action` here. It's not anywhere else in the code.

Reply:
The main point of data-collection-action is to manage remove/replace/append choice automatically for modules.